### PR TITLE
Fix conditional package updating in update all scenarios, add more tests for conditional package operations

### LINKED ISSUE #4681: PM UI multi targeting experience is incomplete - support for updating and uninstalling conditional package versions
If you have a package reference in multiple conditional item groups, then PM UI does not update all occurrences of the package.

Repro Steps- 
1. Start with - 
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>netstandard1.0;netstandard1.3;net45;uap10.0;portable-win81+wpa81;win8;wpa81;wp8</TargetFrameworks>
  </PropertyGroup>
</Project>
```

2. Install NUnit v3.4.0 - 

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>netstandard1.0;netstandard1.3;net45;uap10.0;portable-win81+wpa81;win8;wpa81;wp8</TargetFrameworks>
  </PropertyGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'win8'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'wp8'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'uap10.0'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'wpa81'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'net45'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'portable-win81+wpa81'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
</Project>
```

3. Update NUnit to v3.6.0 through the PM UI - 

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>netstandard1.0;netstandard1.3;net45;uap10.0;portable-win81+wpa81;win8;wpa81;wp8</TargetFrameworks>
  </PropertyGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'win8'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'wp8'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'uap10.0'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'wpa81'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'net45'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'portable-win81+wpa81'">
    <PackageReference Include="NUnit">
      <Version>3.4.0</Version>
    </PackageReference>
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="NUnit" Version="3.6.0" />
  </ItemGroup>
</Project>
```

After the third step, we add an unconditional reference to NUnit v3.6.0 without removing existing conditional package references. This leads to a customer bug - vsfeedback/812089

Related: https://github.com/NuGet/Home/issues/5521

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/4681

Regression? Last working version:

## Description

Conditional package updating was previously fixed in https://github.com/NuGet/NuGet.Client/commit/fed7892af6c079015431389d7cf7a7f630e22bbd. 

This PR fixes the `update all` scenario, ensuring all packages are updated instead of just 1. 
This PR also improves the tests cases for the scenarios.

Some technical details to help with the review of this PR: 

- BuildIntegratedProjectAction is always the result for a project action on a PackageReference project, even if there are multiple packages being updated. Note that this is not something we're adding in this PR. 
- For each package being updated as part of an action, the compatible frameworks should be consider with that package id in mind. That allows, Newtonsoft.Json to be updated for net472, but NuGet.Common to be updated for net5.0.
- BuildIntegratedProjectAction types are not expected to be used outside of the NuGet.PackageManagement, and given how overexposed are APIs are, I'm adding these as internal. This is probably not  the best design, but it avoids adding future work for us. An alternative would be to complete break the type.
- Added nullable annotation for BuildIntegratedProjectAction 

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added, will also add manual tests.
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs]: I think the normal convention is that the license header stays at the top of the file, so this compiler directive should probably be moved down.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3092]: This looks like it'll cause problems when TFMs as aliases gets implemented. Can framework name and target alias be swapped, so the alias is the key, not the value?

Looks like `BuildIntegratedInstallationContext` already exists, not created by this PR, and already uses NuGetFramework as the dictionary key, so my request is out of scope for this PR. 😭

**Review by zivkan — DISMISSED** [review]

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3092]: yeah it's one of many things that'll need tweaked :)

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs]: Fixed, ready for rereview.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:67]: Just to clarify, this [constructor](https://learn.microsoft.com/dotnet/api/system.obsoleteattribute.-ctor?view=net-7.0#system-obsoleteattribute-ctor(system-string)) triggers a warning instead of an error. If we want to raise an error, then we need to use a different constructor.
```cs
[Obsolete("This type is not expected to be created externally.", error: true)]
```

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:165]: I think we should localize this exception message.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3103]: Should we localize this exception message?

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:67]: that would make it a breaking change. I mean, warnings are too if customers use treat warnings as errors, but at least in that case they're shotting themselves in the foot. There have been customers asking us questions about the NuGet.PackageManagement package before, so we know that some people are using it to install packages in their own apps.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3103]: IMO, this one won't benefit. The switch statement already exhausts every defined `NuGetProjectActionType` enum value, but because it's valid, from a compiler's point of view, to do `(NuGetProjectActionType)1234`, the compiler forces you to still have a default value, just in case. But that would be a programming error, one that we wouldn't expect a customer to ever see.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:165]: This is an internal constructor. 
If this happens, it's our mistake.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:176]: I'm really curious to understand why we should access the obsolete members here. We already assigned the new property `ActionAndContextList` with the correct value in this constructor, which isn't obsolete. And have we done the same in the constructor that is marked as obsolete in this PR.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3103]: Yep, these are exceptions that basically tell us that we haven't implemented all the scenarios correctly.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:176]: The constructor is new, there's a new property, but that  doesn't mean that the old contract should be broken. 
Normally, if this wasn't a public API, I'd remove the old properties, but that's not possible without a breaking change currently. 
So until then, we make sure that the type is correctly populated.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:67]: I agree with @zivkan. 

The error option is a really big hammer. We haven't really gotten a lot of feedback on our deprecations to change our approach imo.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:176]: The two properties `OriginalActions` and `InstallationContext` are currently unused in the product code. If a caller uses the new constructor, we'll assign a value to the newly introduced property in this PR, `ActionAndContextList`, which the product code references. However, if the caller uses the old constructor, we'll generate the `ActionAndContextList` property using `OriginalActions` and `InstallationContext`. This new property will then be used throughout the rest of the product code. Therefore, I'm uncertain about the necessity of assigning values to the old or obsolete properties in the new constructor.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:82]: How about modifying the obsolete constructor to call the new one? This approach would allow us to centralize the null checks in one place.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs
Comment (kartheekp-ms) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs:120]: Why did we choose net472 as the unsuccessful framework for this test?

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs
Comment (kartheekp-ms) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs:43]: Adding a blank line after action would clarify that line 41 is an action followed by assertions in line 43. This comment applies to multiple tests in this class.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/PackageReferenceBasedNuGetPackageManagerTests.cs
Comment (kartheekp-ms) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/PackageReferenceBasedNuGetPackageManagerTests.cs:304]: Is it required to pass the same value for both the parameters?

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs]: Out of scope for this PR but accessing `.Count` is better in terms of performance than `.Any()`.

```cs
                    successfulFrameworks.Count > 0 &&
                    unsuccessfulFrameworks.Count > 0 &&
```

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:176]: We don't know who/how people use our code unfortunately.
If people say we're not worried about that, I can remove it, but I was erring on the cautious side.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:82]: The null checks are different across the constructors. We'd end up with an additional constructor.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3111]: If there are multiple items in the collection `projectActionsAndInstallationContexts`, I am curious to learn why we only access first item in the collection.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs:120]: It's part of the test name. 
A package is installed conditionally, but the update to that conditional package failed, hence net472.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/PackageReferenceBasedNuGetPackageManagerTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/PackageReferenceBasedNuGetPackageManagerTests.cs:304]: They are combined with an uniqueness check so not really relevant. 

I think this part of the code is not the most interesting part of the changes though. It's fairly standard test setup code.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3111]: I'm not changing that code.

The way this was originally implemented was that a single PR update event would lead to a single BuildIntegratedProjectAction, which could mean more than 1 write to the project file through update all and similar commands. 
 
 BuildIntegratedProjectAction extends NuGetProjectAction which requires a single action, so that's why the first one was used.
Changing that is difficult without breaking changes.

**Comment by nkolev92** [comment]
@kartheekp-ms 
Since you're adding comments a considerable time apart, can you please let me know when you're done adding comments so I can address the feedback in a single commit?

Note that I've marked comments where my next commit will resolve the feedback as `Resolved` already.

**Comment by kartheekp-ms** [comment]
I am also wondering if we need a new (or update existing) apex test to assert the new behavior.

**Comment by kartheekp-ms** [comment]
> @kartheekp-ms Since you're adding comments a considerable time apart, can you please let me know when you're done adding comments so I can address the feedback in a single commit?
> 
> Note that I've marked comments where my next commit will resolve the feedback as `Resolved` already.

I think I am done for now. Please feel free to push new commit(s).

May be out of scope for this PR. If I click on "Update" a package version in the "Installed" tab, then the versions in the csproj are not updated correctly when there are multiple target frameworks. I can open a separate bug for this scenario.

**Comment by nkolev92** [comment]
> I am also wondering if we need a new (or update existing) apex test to assert the new behavior.

There's no multi targeted templates, so that makes Apex tests extremely difficult to add. .NET templates in previews also require constant source updates right now as packages are not always published to nuget.org, or a source that we control. Created an issue for that https://github.com/NuGet/Client.Engineering/issues/2589.
I'll ask the vendors to add manual tests once the whole thing is complete.

**Comment by nkolev92** [comment]
> May be out of scope for this PR. If I click on "Update" a package version in the "Installed" tab, then the versions in the csproj are not updated correctly when there are multiple target frameworks. I can open a separate bug for this scenario.

With my changes? 
It'd be great if you can give me a setup. 

If you want to consider it out scope, you can just create an issue and assign it to me, and I can validate that afterwards.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3219]: I find it easier to debug separate statements, and I don't expect it's much compilation benefit to inline here.

**Review by donnie-msft — APPROVED** [review]

**Comment by kartheekp-ms** [comment]
> With my changes?
> It'd be great if you can give me a setup.

Yes Nikolche. Here are the repro steps.
 
- Clone this PR branch locally
- Build the repo
- Launch the experimental instance
- Create a new project
- Edit the csproj manually
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>net472;net8.0</TargetFrameworks>  
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>
    <ItemGroup Condition="'$(TargetFramework)' == 'net472'">
	    <PackageReference Include="Newtonsoft.Json">
		    <Version>13.0.1</Version>
	    </PackageReference>
    </ItemGroup>
    <ItemGroup Condition="'$(TargetFramework)' == 'net8.0'">
	    <PackageReference Include="Newtonsoft.Json">
		    <Version>13.0.2</Version>
	    </PackageReference>
    </ItemGroup>
</Project>
```
- Open NuGet Package Manager UI for the project
- Navigate to "Installed" tab
- Click on the update icon with text 13.0.3
image
- As you can see in the below screenshot, package version changed only for "net472" target framework.
image

> If you want to consider it out scope, you can just create an issue and assign it to me, and I can validate that afterwards.

Please take a look at the repro steps and let me know your feedback so that we can plan accordingly.

**Comment by nkolev92** [comment]
@kartheekp-ms 
Turns out project-system is having issues with where the condition is: 

For example, I tested a similar scenario as yours, but the conditions are on the package reference item instead, and that does work!

repro


I'll create an issue for follow up. That change may not be on the nuget side.

**Comment by nkolev92** [comment]
Actually, nevermind you have 2 frameworks in your test and I have 3. 
The problem is that we don't consider the fact that all frameworks have a conditional reference. 

I'll create a follow up for that.
That *may* be something we can detect, but I'm not sure we can have a perfect solution.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3219]: This is fairly easy to debug if you put a breakpoint outside of the method where you'd get the already created installation context. 
It's not really for a compilation benefit, I just thought this was readable, and there was no need to add extra lines.

**Comment by kartheekp-ms** [comment]
> I'll create a follow up for that.
> That may be something we can detect, but I'm not sure we can have a perfect solution.

Has the follow-up issue been created?

**Comment by nkolev92** [comment]
No, I was waiting for approval on this first. 

I created it now https://github.com/NuGet/Home/issues/13034

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:82]: Can we extract all the common null checks into another method and invoke that method from the constructor?

Here's a suggested implementation of the `ValidateArguments` method:

```cs
private void ValidateArguments(NuGetProject project, PackageIdentity packageIdentity, LockFile originalLockFile, RestoreResultPair restoreResultPair, IReadOnlyList<SourceRepository> sources, object actions, BuildIntegratedInstallationContext? installationContext = null)
{
    if (project == null)
        throw new ArgumentNullException(nameof(project));
    if (originalLockFile == null)
        throw new ArgumentNullException(nameof(originalLockFile));
    if (restoreResultPair == null)
        throw new ArgumentNullException(nameof(restoreResultPair));
    if (sources == null)
        throw new ArgumentNullException(nameof(sources));
}
```

We can add null checks that are specific to a particular constructor within that constructor itself, after validating the common arguments.

We can remove the following null check from the constructors in this class because the [base constructor takes care of that](https://github.com/NuGet/NuGet.Client/blob/d85c6ea4588c5b4f2b4a4d7bf8073837fb308f58/src/NuGet.Core/NuGet.PackageManagement/NuGetProjectAction.cs#L49-L52):

```cs
if (packageIdentity == null)
    throw new ArgumentNullException(nameof(packageIdentity));
```

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3111]: Do you think there is any technical debt in this code path that requires a closer look at the API? If yes, feel free to create a new technical debt issue.

> The way this was originally implemented was that a single PR update event would lead to a single BuildIntegratedProjectAction, which could mean more than 1 write to the project file through update all and similar commands.

Will there be any consistency issues if there are multiple writes to the same project file?

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs
Comment (kartheekp-ms) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs:120]: Thanks for the clarification; I now understand. Essentially, the test is to verify that if an update fails for the `net472` framework, then the collection of unsuccessful frameworks should include both `net5`, since it lacks dependencies on package 'a', and 'net472', because the update failed for some reason. Is my understanding correct?

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/PackageReferenceBasedNuGetPackageManagerTests.cs
Comment (kartheekp-ms) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/PackageReferenceBasedNuGetPackageManagerTests.cs:304]: I agree with your feedback. Maybe we could have passed an empty list (Enumerable.Empty<SourceRepository>) because there are no secondary sources in this test.

**Review by kartheekp-ms — APPROVED** [review]
There are some unresolved comments in this pull request, primarily the ones we discussed earlier. Feel free to review them, but I don't want to block this PR until those comments are addressed.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerUtilityTests.cs:120]: yes

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/PackageReferenceBasedNuGetPackageManagerTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/PackageReferenceBasedNuGetPackageManagerTests.cs:304]: Nearly all of our tests do the same thing, so I don't think it warrants a change. This is what the PM UI would pass in for PR anyways.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:3111]: > Do you think there is any technical debt in this code path that requires a closer look at the API? If yes, feel free to create a new technical debt issue.

Maybe?

1 action meaning more packages has permeated throughout the APIs and I'm not really sure if that's necessarily the best design, but it's also not a frequent code path, but the public APIs just make things nearly impossible to change.
I'm gonna look into the whole tech debt thing, and I'll consider this then.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/BuildIntegratedProjectAction.cs:82]: I can clean that up in a future PR.

**Pull request merged**
